File: LinkedList/LinkedList.cpp
#include <iostream>

struct Node {
  int key;
  Node * next;
  Node(int k){
    key = k;
    next = NULL;
  }
};

Node* insertBegin(Node* head, int k){
  Node* temp = new Node(k);
  temp->next = head;
  return temp;
}

Node* insertEnd(Node* head, int k){
  if(head== NULL) return new Node(k);
  Node *curr = head;
  while(curr->next != NULL){
    curr= curr->next;
  }
  curr->next = new Node(k);
  return head;
}
//1 2 3 4 5 6
Node* middleNode(Node* head){
  Node* fast = head;
  Node* slow = head;
  while(fast!= NULL && fast->next != NULL){
    fast = fast->next->next;
    slow = slow->next;
  }
  return slow;
}

Node* nthNodeFromEnd(Node* head,int n){
  Node* first= head;
  Node* second = head;
  while(n>0){
    first = first->next;
    n--;
  }
  while(first != NULL){
    first = first->next;
    second = second->next;
  }
  return second;
}
void printList(Node *head){
  // loop ->next until you reach head eq NULL
  while(head != NULL){
    std::cout <<head->key<<" ";
    head = head->next;
  }
}

Node* reverseList(Node* head){
  Node* curr = head;
  Node* prev = NULL;
  Node* next = NULL;
  while(curr!= NULL){
    next = curr->next;
    curr->next = prev;
    prev = curr;
    curr = next;
  }
  return prev;
}

int main() {
  Node* list ;
  // list = insertBegin(list, 24);
  // list = insertBegin(list, 36);
  list = insertEnd(list, 10);
  list = insertEnd(list, 20);
  list = insertEnd(list, 30);
  list = insertEnd(list, 40);
  list = insertEnd(list, 50);
  list = insertEnd(list, 60);
  std::cout <<middleNode(list)->key<<"  __ \n";
  std::cout <<nthNodeFromEnd(list, 3)->key<<"  __ \n";
  list = reverseList(list);
  printList(list);
}

File: LinkedList/LinkedList.java
class Node{
  int data;
  Node next;
  Node(int key){
    this.data = key;
  }
}

class List{
  Node head;
  void insertBegin(int key){
    Node temp = new Node(key);
    temp.next = head;
    head = temp;
  }
  void insertLast(int key){
    if(head == null){
      head = new Node(key);
      return;
    }
    Node temp = head;
    while(temp.next != null){
      temp = temp.next;
    }
    Node newNode = new Node(key);
    temp.next = newNode;
  }
  int middleNode(){
    Node slow = head;
    Node fast = head;
    while(fast != null && fast.next != null){
      slow = slow.next;
      fast = fast.next.next;
    }
    System.out.println("mid:  "+slow.data);
    return 0;
  }

  Node nthNodeFromEnd(int n){
    Node first = head;
    Node second = head;
    while(n>0){
      second = second.next;
      n--;
    }
    while(second != null){
      if(first == null) return null;
      first = first.next;
      second = second.next;
    }
    return first;
  }
  
  void printList(){
    Node temp = head;
    while(temp != null){
      System.out.print(temp.data+" ");
      temp = temp.next;
    }
  }
  void reverseList(){
    Node curr = head;
    Node prev = null;
    Node next;
    while(curr!=null){
      next = curr.next;
      curr.next = prev;
      prev = curr;
      curr = next;
    }
    head = prev;
  }
}
class Main {
  public static void main(String[] args) {
    List list = new List();
    list.insertLast(10);
    list.insertLast(20);
    list.insertLast(30);
    list.insertLast(40);
    list.insertLast(50);
    list.insertLast(60);
    list.middleNode();
    list.printList();
    System.out.println("\nNth node from end: "+list.nthNodeFromEnd(4).data);
    list.reverseList();
    list.printList();
  }
}